### PR TITLE
[xbmc][win32] First step in supporting VS2017

### PR DESCRIPTION
**Pull request by Paxxi**
## Description

Adds vswhere as a package.

Will try to find vs2017 and fall back to vs2015 in case it's not found.
SDK version is still hard coded and pinned to 14393 to avoid any breakage on jenkins.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: tools/buildsteps/windows/BuildSetup.bat [hunk 1]
```diff
@@ -1,319 +1,319 @@
...
+PACKAGE_IDENTITY ^
+PACKAGE_PUBLISHER ^
+VERSION_MAJOR ^
+VERSION_MINOR ^
+VERSION_TAG ^
+VERSION_CODE ^
+WEBSITE
+
+FOR %%p IN (%version_props%) DO (
+  FOR /f "delims=" %%v IN ('%sed_exe% -n "/%%p/ s/%%p *//p" %base_dir%\version.txt') DO SET %%p=%%v
```
Comment (afedchin) [inline tools/buildsteps/windows/BuildSetup.bat]: `^M` shouldn't be here

**Review by afedchin — APPROVED** [review]
looks good in general, but something wrong with BuildSetup.bat

**Comment by Rechi** [comment]
@Paxxi fix the commit messages and I'll review it afterwards

File: tools/buildsteps/windows/BuildSetup.bat (hunk 1 above)
Comment (Paxxi) [reply tools/buildsteps/windows/BuildSetup.bat]: yeah I think it got a bit messed up because I had turned off autocrlf in git

Comment (Paxxi) [reply tools/buildsteps/windows/BuildSetup.bat]: hmm I've squashed it and looked at the end of line characters for everything and this line has the same crlf as every other line but for some reason github shows the ^M character.

**Comment by Paxxi** [comment]
@Rechi squashed it now and gotten rid of all the nonsense commits

**Comment by Rechi** [comment]
whole BuildSetup.bat file still has a line ending change

**Comment by Paxxi** [comment]
ok third time's the charm :)

**Comment by Rechi** [comment]
The 64 bit toolchains should still be the preferred ones, but with your changes they aren't.
WIN-UWP-ARM fails, because you have to call vcvarsall.bat with `x86_arm` or `amd64_arm` if you run it on a non ARM OS.

**Comment by Paxxi** [comment]
ahh that's the issue with arm. I'll look at fixing those bits.
Does it look ok otherwise?

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by Paxxi** [comment]
So here's the plan for the upgrade to vs2017.

1. Get this merged
2. Install vs2017 and sdk ver 10.0.16299.0 on all build machines
3. PR a bump to the sdk version
4. uninstall vs2015 and sdk 10.0.14393.0 on all build machines
5. profit

**Comment by afedchin** [comment]
should we update vcredist?

**Comment by Paxxi** [comment]
forgot about vcredist. Not sure if we need to but makes sense to bump it as well to match.

File: .gitignore [hunk 2]
```diff
@@ -58,7 +59,7 @@ testMain
 *.diagsession
 
 # osx specifics
-/build/
+/build*/
```
Comment (Rechi) [inline .gitignore]: should go into a separate commit, if you want to keep it

File: tools/buildsteps/windows/BuildSetup.bat [hunk 3]
```diff
@@ -1,5 +1,5 @@
 @ECHO OFF
-SETLOCAL ENABLEDELAYEDEXPANSION
+REM SETLOCAL ENABLEDELAYEDEXPANSION
```
Comment (Rechi) [inline tools/buildsteps/windows/BuildSetup.bat]: If not needed remove it, instead commenting it

File: tools/buildsteps/windows/vswhere.bat [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+SET vcarch=amd64
+SET vcstore=%2
+SET vcvars=no
+SET sdkver=
+
+SET vsver=
+SET toolsdir=%arch%
+
+IF "%arch%" NEQ "x64" (
+  SET vcarch=%vcarch%_%arch%
```
Comment (Rechi) [inline tools/buildsteps/windows/vswhere.bat]: We should try with `amd64_%arch%` first and if that fails fall back to `%arch%` like we currently do it.

File: tools/buildsteps/windows/vswhere.bat [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    SET vcvars="%%i\VC\Auxiliary\Build\vcvarsall.bat"
+    SET vsver=15
+  )
+)
+
+IF %vcvars%==no (
+  FOR /f "usebackq tokens=1* delims=" %%i in (`%vswhere% -legacy -property installationPath`) do (
+    IF EXIST "%%i\VC\vcvarsall.bat" (
+      SET vcvars="%%i\VC\vcvarsall.bat"
+      SET vsver=14
```
Comment (Rechi) [inline tools/buildsteps/windows/vswhere.bat]: exit for loop after first existing path

File: tools/buildsteps/windows/win32-uwp/make-addons.bat [hunk 6]
```diff
@@ -1,8 +1,6 @@
-@ECHO OFF
```
Comment (Rechi) [inline tools/buildsteps/windows/win32-uwp/make-addons.bat]: line endings changed in whole file

**Review by Rechi — COMMENTED** [review]
Should we already use VS2017 as preferred version and use VS2015 as fallback?

File: tools/buildsteps/windows/vswhere.bat (hunk 5 above)
Comment (Paxxi) [reply tools/buildsteps/windows/vswhere.bat]: it's doesn't really matter much as for most systems we'll only get a single result back

File: tools/buildsteps/windows/vswhere.bat (hunk 4 above)
Comment (Paxxi) [reply tools/buildsteps/windows/vswhere.bat]: do we really care about building on 32-bit versions?

File: tools/buildsteps/windows/vswhere.bat (hunk 5 above)
Comment (Rechi) [reply tools/buildsteps/windows/vswhere.bat]: On my system I get two results back, although for the second one (12.0) the if check returns false but others may have installed older versions of VS.

File: tools/buildsteps/windows/vswhere.bat (hunk 4 above)
Comment (Rechi) [reply tools/buildsteps/windows/vswhere.bat]: Will the `amd64_%arch%` version always be available if `%arch% is available?
If yes I don't care much, otherwise it has to be adjusted.

File: tools/buildsteps/windows/win32-uwp/make-addons.bat (hunk 6 above)
Comment (Paxxi) [reply tools/buildsteps/windows/win32-uwp/make-addons.bat]: Not entirely sure how to fix that. I've since switched my checked out repo to autocrlf=true and cloning a new version without it checks it out as LF file endings.

File: tools/buildsteps/windows/vswhere.bat (hunk 5 above)
Comment (Paxxi) [reply tools/buildsteps/windows/vswhere.bat]: I was thinking of the wrong check. This check will be removed after we bump the sdk, it's only there currently to not break jenkins so we can do the upgrade in steps.

Also it seems harder than it should be to break out of a for loop see https://stackoverflow.com/questions/6728798/exiting-out-of-a-for-loop-in-a-batch-file haven't tested it myself though.

File: tools/buildsteps/windows/vswhere.bat (hunk 4 above)
Comment (Paxxi) [reply tools/buildsteps/windows/vswhere.bat]: Should be. There's no separate option to install one or the other
image

**Comment by Paxxi** [comment]
I think we should as this is intended to upgrade the vs2017 and build hosts don't have it installed it's a safe setting imho.

Dropped the gitignore change and remove the commented out setlocal. Seem to have fixed the line endings but I'm never really sure  😄

**Comment by Paxxi** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by Rechi** [comment]
has to be squashed

**Review by Rechi — APPROVED** [review]

**Comment by MartijnKaijser** [comment]
jenkins build this please
